src/System.Collections.Concurrent/tests/Performance/Perf.ConcurrentBag.cs

```diff
@@ -0,0 +1,192 @@
+
+        [Benchmark(InnerIterationCount = 10_000_000), MeasureGCAllocations]
+        public void Add()
+        {
+            foreach (BenchmarkIteration iteration in Benchmark.Iterations)
+            {
+                long size = Benchmark.InnerIterationCount;
+                var q = new ConcurrentBag<int>();
+
+                using (iteration.StartMeasurement())
+                {
+                    for (int i = 0; i < size; i++)
+                    {
+                        q.Add(i);
+                    }
+                }
+            }
+        }
+
+        [Benchmark(InnerIterationCount = 1_000_000), MeasureGCAllocations]
+        [InlineData(false)]
+        [InlineData(true)]
+        public void TryTake(bool exists)
+        {
+            foreach (BenchmarkIteration iteration in Benchmark.Iterations)
+            {
+                long size = Benchmark.InnerIterationCount;
+                var b = new ConcurrentBag<int>();
+
+                if (exists)
+                {
+                    for (int i = 0; i < size; i++)
+                    {
+                        b.Add(i);
+                    }
+                }
+
+                using (iteration.StartMeasurement())
+                {
+                    for (int i = 0; i < size; i++)
+                    {
+                        b.TryTake(out int result);
+                    }
+                }
+            }
+        }
+
+        [Benchmark(InnerIterationCount = 1_000_000), MeasureGCAllocations]
+        public void AddTake()
+        {
+            foreach (BenchmarkIteration iteration in Benchmark.Iterations)
+            {
+                long size = Benchmark.InnerIterationCount;
+                var b = new ConcurrentBag<int>();
+
+                using (iteration.StartMeasurement())
+                {
+                    for (int i = 0; i < size; i++)
+                    {
+                        b.Add(i);
+                        b.TryTake(out int result);
+                    }
+                }
+            }
+        }
+
+        [Benchmark(InnerIterationCount = 1_000_000), MeasureGCAllocations]
+        [InlineData(0)]
+        [InlineData(3)]
+        public void Pooling_AddTakeFromSameThreads(int initialSize)
+        {
+            foreach (BenchmarkIteration iteration in Benchmark.Iterations)
+            {
+                const int NumThreads = 2;
+                long size = Benchmark.InnerIterationCount;
+                var b = new ConcurrentBag<int>();
+
+                using (var barrier = new Barrier(NumThreads + 1))
+                {
+                    Task[] tasks = Enumerable.Range(0, NumThreads).Select(_ => Task.Factory.StartNew(() =>
+                    {
+                        for (int i = 0; i < initialSize; i++)
+                        {
+                            b.Add(i);
+                        }
+
+                        barrier.SignalAndWait();
+                        barrier.SignalAndWait();
```

Comment:
why the two signals?

Reply:
one to correspond to line 135 and one to correspond to line 138?

Reply:
Below we have a SignalAndWait before and after the call to StartMeasurement.  My goal here is to measure the actual work as best as possible, rather than measuring overheads associated with creating/scheduling the threads.  So we get the threads to the point where they're about to do work, then we start the measuring, and then we let them proceed to do the work.